### DISCUSSION TITLE
Make runtime footprint smaller

### DISCUSSION THREAD

**Discussion by rigor789** [NativeScript/rfcs#44] [1. Ideas & Discussions]
Right now if you are working with multiple projects, all of them have a pretty large `node_modules` folder, partially due to the size of the ios and android runtimes, and a lot of storage is wasted and it adds up quickly.

Proposal: make `@nativescript/ios` and `@nativescript/android` lightweight "placeholders" that indicate to the CLI (or a postinstall script / script invoked by the CLI) where/how to fetch the runtime files and reuse them if they have already been downloaded to a global cache directory.

- make npm packages lightweight
- cache runtime files (per version) in a global cache folder
- perform checksums to avoid corrupted downloads/caches
- reuse cached runtime if present
- keep a "last-used" metadata file in cache directory and auto-cleanup when a specific version has not been used for X amount of time (prevents excessive cache buildup over time) 
- allow cleaning the cache easily - perhaps `ns clean --full` or something similar

The actual runtime files would need to be hosted perhaps as GitHub release assets, this definitely complicates the publish pipeline, however we are (I believe) getting close to the npm package size limit, so sooner or later we might need to do this anyways.

/cc @edusperoni @triniwiz @NathanWalker

**Comment by farfromrefug** [comment]
I really like the idea. Implementing that we could also detect runtime version change and force a clean of platforms in this case. Could prevent a lot of false positive bug reports.

**Comment by edusperoni** [comment]
I've mentioned this before, but prisma is a good example of this.

https://www.npmjs.com/package/@prisma/engines is responsible for downloading and setting up the versions. They use https://binaries.prisma.sh/

They also allow you to specify specific files for it to use, so if you want to test some other binary version you just need to specify some environment variables

**Reply by rigor789** [reply]
The only thing to consider is hosting costs - that's where GitHub release artifacts are nice - and pretty sure they can be automated too!

**Reply by farfromrefug** [reply]
@rigor789 they definitely can. I already build android v8 through github action and adding them to a release is really easy

**Comment by ammarahm-ed** [comment]
I think a simpler solution is to publish ios runtime on Pods & android runtime on maven. This makes npm installs small & gradle/cocoapods hande all the painful stuff for us automatically when you build the native app.

**Comment by farfromrefug** [comment]
@ammarahm-ed I dont think pods can actually.host any file. Pod repo only holds manifests. The file needs to be hosted somewhere else. 
Ans maven might have a size limit.

**Reply by rigor789** [reply]
We are experimenting with hosting the large files on GitHub Releases/Artifacts and linking to those from a Podspec (for iOS).

If that proves to work nicely, we can do something similar for Android (tbd)

/cc @ammarahm-ed

**Reply by ammarahm-ed** [reply]
Awesome. I think it will work out nicely. I have seen many projects use it and they work great.
